### PR TITLE
Implemented functions feof(), ferror() and clearerr()

### LINKED ISSUE #997: Implement ferror and friends
There is a bunch of functions (clearerr, feof, ferror, fileno) in POSIX that are used to check and reset stream status.

### LINKED ISSUE #1746: Function feof() is not yet implemented
The function `feof()` located [here](https://github.com/embox/embox/blob/master/src/compat/libc/stdio/stdio_stubs.c), which is used to detect the EOF (End Of File), is not yet implemented.

### PR DESCRIPTION
**Pull request by MrSquanchee**
closes #997: Implementing ferror and friends and closes #1746
I have added functionality for detection of EOF and any ERR in the
files->flags by creating macros IO_EOF_ (0x0010) and IO_ERR_ (0x0020)
and updating files->flags, whenever any file IO function encounters
EOF or any kind of error occurs, which is detected by functions
feof() and ferror() and cleared by clearerr().
file IO functions where EOF may occur -> fread()
file IO functions where ERR is set    -> fflush(), fread(), fseek()
and fwrite().

### CODE REVIEW

File: src/compat/libc/stdio/file_struct.h [hunk 1]
```diff
@@ -31,4 +31,13 @@ struct file_struct {
 
 extern int funopen_check(FILE *f);
 
+#define IO_EOF_		   		0x0010  /* To check if EOF is seen*/
+#define IO_ERR_		   		0x0020  /* To check if an error is seen*/
+#define SET_IO_EOF(fp)		((fp)->flags |= IO_EOF_) /*Sets the field 0x0010*/
+#define SET_IO_ERR(fp)		((fp)->flags |= IO_ERR_) /*Sets the field 0x0020*/
+#define IO_EOF_DETECT(fp)  	(((fp)->flags & IO_EOF_) != 0 ? 1:0) /*True if 0x0010 is set*/
+#define IO_ERR_DETECT(fp)  	(((fp)->flags & IO_ERR_) != 0 ? 1:0) /*True if 0x0020 is set*/
```
Comment (alexkalmuk) [inline src/compat/libc/stdio/file_struct.h]: Can you please fix formatting?

File: src/compat/libc/stdio/file_struct.h [hunk 2]
```diff
@@ -31,4 +31,13 @@ struct file_struct {
...
+#define IO_CLEARERR(fp)		((fp)->flags &= ~(IO_ERR_)); \
+	((fp)->flags &= ~(IO_EOF_))
```
Comment (alexkalmuk) [inline src/compat/libc/stdio/file_struct.h]: You can just do `(fp)->flags &= (IO_ERR_ | IO_EOF_)`

**Comment by alexkalmuk** [comment]
To reproduce the error with lua you can comment all lines in `conf/start_script.inc`, then rebuild and launch embox. Then,

```
root@embox:(null)#ls fib.lua                                                    
 fib.lua
root@embox:(null)#lua fib.lua                                                   
lua: cannot read fib.lua: No error
lua: Command returned with code 1: Unknown error code. PLEASE REPORT ISSUE 519!
tish error: #1
```

**Comment by alexkalmuk** [comment]
After that set breakpoints to feof, ferror and clearerr and see what happen

**Comment by alexkalmuk** [comment]
As far as I can see something goes wrong when lua call ferror.

**Comment by MrSquanchee** [comment]
I have thought out the changes I have to make in order for the code to behave properly
1. Removed the error checking(s) added in `fflush()` and `fseek()`.
2. Only detecting stream error when a user tries to read from a write only file and write to a read only file
3. I have tested the function `feof()`, `clearerr()` and `ferror()`  which behave properly i.e. detecting and clearing EOF and ERR flags from `file->flags` and clearing it.
4. Now `ferror()` only detects errors produced by 2.

@alexkalmuk . If you approve of these changes I will squash commit and force push the PR branch.

File: src/compat/libc/stdio/fread.c [hunk 3]
```diff
@@ -50,6 +51,13 @@ size_t fread(void *buf, size_t size, size_t count, FILE *file) {
 		/* try to revert some bytes */
 		cnt -= (cnt % size);
 	}
+	
+	if ((file->flags & O_ACCESS_MASK) == O_WRONLY) {
+		SET_IO_ERR(file);
+	}
```
Comment (alexkalmuk) [inline src/compat/libc/stdio/fread.c]: Are you sure we need to check this after reading?

**Comment by alexkalmuk** [comment]
> I have tested the function feof(), clearerr() and ferror() which behave properly i.e. detecting and clearing EOF and ERR flags from file->flags and clearing it.

How?

File: src/compat/libc/stdio/fwrite.c [hunk 4]
```diff
@@ -146,5 +147,9 @@ size_t fwrite(const void *buf, size_t size, size_t count, FILE *file) {
 		}
 	}
 
+	if ((file->flags & O_ACCESS_MASK) == O_RDONLY) {
+		SET_IO_ERR(file);
+	}
```
Comment (alexkalmuk) [inline src/compat/libc/stdio/fwrite.c]: Same here.

**Comment by MrSquanchee** [comment]
> > I have tested the function feof(), clearerr() and ferror() which behave properly i.e. detecting and clearing EOF and ERR flags from file->flags and clearing it.
> 
> How?

I did that using a command locally in my system for when I try to do something like making a test in src/test this happens ....
testingError

**Comment by alexkalmuk** [comment]
> fopen() doesn't seem to work at this point of time. But when I use a command for the tests it turns out fine. Any help would be appreciated.

I do not see the test. You didn't push it?

**Comment by MrSquanchee** [comment]
> I do not see the test. You didn't push it?

I pushed it now.

**Comment by alexkalmuk** [comment]
> > I do not see the test. You didn't push it?
> 
> I pushed it now.

Ok, thank, I will take a look now

**Comment by alexkalmuk** [comment]
@MrSquanchee I added `@Runlevel(2) include embox.test.stdio.fileIOErrorTest` to `x86/qemu` an the test passed.

**Comment by MrSquanchee** [comment]
It means a merge :) or does it ???

File: src/tests/stdio/Mybuild [hunk 5]
```diff
@@ -22,3 +22,14 @@ module fgets_test {
 	depends embox.compat.libc.all
 	depends embox.framework.LibFramework
 }
+
+module fileIOErrorTest {
+	source "fileIOErrorTest.c"
+	
+	@InitFS
+	source  "fileIOErrorTest.txt"
```
Comment (alexkalmuk) [inline src/tests/stdio/Mybuild]: Our code styles assume underscores in function and files namings -https://github.com/embox/embox/wiki/Code-Style

Can you please fix this?

**Comment by alexkalmuk** [comment]
> @alexkalmuk It means a merge :) or does it ???

I was wondering why you told `fopen` didn't open your file, that's why I tried your changes. Yes, please module namings and we can merge it.

**Comment by MrSquanchee** [comment]
@alexkalmuk I guess I was doing something wrong .... 
Thanks for testing it :)

File: src/tests/stdio/file_io_error_test.c [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+ * @author Suraj Upadhyay
+ * @date 08.03.2020
+ */
+
+#include <stdio.h>
+#include <embox/test.h>
+
+EMBOX_TEST_SUITE("stdio/feof ferror clearerr test");
+
+#define TEST_FILE_NAME "/fileIOErrorTest.txt"
```
Comment (alexkalmuk) [inline src/tests/stdio/file_io_error_test.c]: Are you sure the still works?

Comment (alexkalmuk) [reply src/tests/stdio/file_io_error_test.c]: Because you renamed the file to file_io_error_test.txt

Comment (MrSquanchee) [reply src/tests/stdio/file_io_error_test.c]: A lot of oopsies these days .... 
I will change it shortly :)

**Comment by MrSquanchee** [comment]
@alexkalmuk . I just realized that we can delete this [stdio_stubs.c](https://github.com/embox/embox/blob/master/src/compat/libc/stdio/stdio_stubs.c) now.

**Comment by alexkalmuk** [comment]
> @alexkalmuk . I just realized that we can delete this [stdio_stubs.c](https://github.com/embox/embox/blob/master/src/compat/libc/stdio/stdio_stubs.c) now.

Yes, I agree, let's remove it.

**Comment by MrSquanchee** [comment]
> > @alexkalmuk . I just realized that we can delete this [stdio_stubs.c](https://github.com/embox/embox/blob/master/src/compat/libc/stdio/stdio_stubs.c) now.
> 
> Yes, I agree, let's remove it.

@alexkalmuk I deleted stdio_stubs.c. I think now we should be able to merge.
Wdyt :)

File: src/tests/stdio/Mybuild [hunk 7]
```diff
@@ -22,3 +22,14 @@ module fgets_test {
 	depends embox.compat.libc.all
 	depends embox.framework.LibFramework
 }
+
+module fileIOErrorTest {
```
Comment (alexkalmuk) [inline src/tests/stdio/Mybuild]: the module name is still without underscores.

**Comment by alexkalmuk** [comment]
@MrSquanchee 

```diff
diff --git a/templates/x86/qemu/mods.config b/templates/x86/qemu/mods.config
index f98e665..5197c72 100644
--- a/templates/x86/qemu/mods.config
+++ b/templates/x86/qemu/mods.config
@@ -81,6 +81,8 @@ configuration conf {
        @Runlevel(2) include embox.test.critical
        @Runlevel(2) include embox.test.recursion
 
+       @Runlevel(2) include embox.test.stdio.fileIOErrorTest
+
        @Runlevel(2) include embox.test.stdio.printf_test
        @Runlevel(2) include embox.test.posix.poll_test
        @Runlevel(2) include embox.test.posix.select_test
```

Then make and run:

```
failure at src/tests/stdio/file_io_error_test.c : 27, in function __test_case_at_line_17
		test_assert_equal(fgetc(f), EOF)
	   case at src/tests/stdio/file_io_error_test.c : 17
		"feof and clearerr should set and clear EOF flag"
	.
	testing fileIOErrorTest (stdio/feof ferror clearerr test) failed
		1/2 failures
```

**Comment by MrSquanchee** [comment]
@alexkalmuk I think that's because the newline introduced after `a`.

**Comment by alexkalmuk** [comment]
> @alexkalmuk I think that's because the newline introduced after `a`.

Hm, that makes sense, could you fix this?

**Comment by MrSquanchee** [comment]
@alexkalmuk . This time I won't ask you. You are damn good at your job :)

**Comment by alexkalmuk** [comment]
Finally, the test works, thank you!

**Comment by alexkalmuk** [comment]
> @alexkalmuk . This time I won't ask you. You are damn good at your job :)

You are free to ask questions, it's normal.

**Pull request merged**